pkg/traversal/traversal.go

```diff
@@ -0,0 +1,293 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package traversal
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/ethersphere/bee/pkg/collection/entry"
+	"github.com/ethersphere/bee/pkg/file"
+	"github.com/ethersphere/bee/pkg/file/joiner"
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/manifest"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+// Service is the service to find dependent chunks for an address.
+type Service interface {
+	// TraverseChunkAddresses iterates through each address related to
+	// the supplied one, if possible.
+	TraverseChunkAddresses(context.Context, swarm.Address, bool, swarm.AddressIterFunc) error
+}
+
+type traversalService struct {
+	logger logging.Logger
+	storer storage.Storer
+}
+
+func NewService(logger logging.Logger, storer storage.Storer) (Service, error) {
+	return &traversalService{
+		logger: logger,
+		storer: storer,
+	}, nil
+}
+
+func (s *traversalService) TraverseChunkAddresses(
+	ctx context.Context,
+	reference swarm.Address,
+	encrypted bool,
+	chunkAddressFunc swarm.AddressIterFunc,
+) error {
+
+	isFile, e, metadata, err := s.checkIsFile(ctx, reference)
+	if err != nil {
+		return err
+	}
+
+	// reference address could be missrepresented as file when:
+	// - content size is 64 bytes (or 128 for encrypted reference)
+	// - second reference exists and is JSON (and not actually file metadata)
+
+	if isFile {
+
+		isManifest, m, err := s.checkIsManifest(ctx, reference, encrypted, e, metadata)
+		if err != nil {
+			return err
+		}
+
+		// reference address could be missrepresented as manifest when:
+		// - file content type is actually on of manifest type (manually set)
+		// - content was unmarshalled
+		//
+		// even though content could be unmarshaled in some case, iteration
+		// through addresses will not be possible
+
+		if isManifest {
+			// process as manifest
+
+			err = m.IterateAddresses(ctx, func(manifestNodeAddr swarm.Address) (stop bool) {
+				err := s.traverseChunkAddressesFromManifest(ctx, manifestNodeAddr, encrypted, chunkAddressFunc)
+				if err != nil {
+					stop = true
+				}
+				return
+			})
+			if err != nil {
+				return fmt.Errorf("traversal: iterate chunks: %s: %w", reference, err)
+			}
+
+			metadataReference := e.Metadata()
+
+			err = s.processBytes(ctx, metadataReference, chunkAddressFunc)
+			if err != nil {
+				return nil
+			}
+
+			if stop := chunkAddressFunc(reference); stop {
```

Comment:
it seems not to matter if stop is true you return nil anyways

Reply:
Yeah, I had several iterations of various stuff, and changed orders on some places, which is why I had some leftovers. Thanks for comments.